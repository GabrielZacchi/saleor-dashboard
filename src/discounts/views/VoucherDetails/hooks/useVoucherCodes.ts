import useListSettings from "@dashboard/hooks/useListSettings";
import { ListViews } from "@dashboard/types";
import { useState } from "react";

import { getVoucherCodesToDisplay } from "../utils";
import { useVoucherCodesClient } from "./useVoucherCodesClient";
import { useVoucherCodesRowSelection } from "./useVoucherCodesRowSelection";
import { useVoucherCodesServer } from "./useVoucherCodesServer";

export const useVoucherCodes = ({ id }: { id: string }) => {
  const {
    settings: voucherCodesSettings,
    updateListSettings: updateVoucherCodesListSettings,
  } = useListSettings(ListViews.VOUCHER_CODES);

  const [isServerPagination, setIsServerPagination] = useState(true);

  const {
    clientVoucherCodes,
    clientVoucherCodesPagination,
    freeSlotsInClientPagianationPage,
    handleAddVoucherCode,
    handleGenerateMultipeCodes,
    hasClientPaginationNextPage,
    hasClientPaginationPrevPage,
    onSettingsChange,
  } = useVoucherCodesClient(voucherCodesSettings, () => {
    setIsServerPagination(false);
  });

  const {
    freeSlotsInServerPagianationPage,
    hasServerPaginationNextPage,
    hasServerPaginationPrevPage,
    serverVoucherCodesPagination,
    serverVoucherCodes,
    voucherCodesLoading,
    voucherCodesRefetch,
  } = useVoucherCodesServer({
    id,
    settings: voucherCodesSettings,
    isServerPagination,
    paginationState: {
      first:
        !isServerPagination && freeSlotsInClientPagianationPage > 0
          ? freeSlotsInClientPagianationPage
          : voucherCodesSettings.rowNumber,
    },
  });

  const voucherCodes = getVoucherCodesToDisplay({
    clientVoucherCodes,
    freeSlotsInClientPagianationPage,
    hasClientPaginationNextPage,
    freeSlotsInServerPagianationPage,
    hasServerPaginationPrevPage,
    isServerPagination,
    serverVoucherCodes,
  });

  const voucherCodesPagination = isServerPagination
    ? serverVoucherCodesPagination
    : clientVoucherCodesPagination;

  const { selectedVoucherCodesIds, handleSetSelectedVoucherCodesIds } =
    useVoucherCodesRowSelection(voucherCodes);

  const handleLoadNextPage = () => {
    if (!hasClientPaginationNextPage) {
      setIsServerPagination(true);
    }

    if (isServerPagination) {
      serverVoucherCodesPagination.loadNextPage();
    } else if (!isServerPagination && freeSlotsInClientPagianationPage > 0) {
      serverVoucherCodesPagination.loadNextPage();
    }

    clientVoucherCodesPagination.loadNextPage();
  };

  const handleLoadPrevousPage = () => {
    if (!hasServerPaginationPrevPage) {
      setIsServerPagination(false);
    }

    if (isServerPagination && hasServerPaginationPrevPage) {
      serverVoucherCodesPagination.loadPreviousPage();
    }

    clientVoucherCodesPagination.loadPreviousPage();
  };

  const calculateHasNextPage = () => {
    // In case when client voucher codes takes all slots
    // on page and there are some server voucher codes to display
    if (
      !isServerPagination &&
      !hasClientPaginationNextPage &&
      freeSlotsInClientPagianationPage === 0 &&
      serverVoucherCodes.length > 0
    ) {
      return true;
    }

    return hasClientPaginationNextPage || hasServerPaginationNextPage;
  };

  const calculateHasPrevPage = () => {
    if (isServerPagination) {
      return hasServerPaginationPrevPage || hasClientPaginationPrevPage;
    }

    return hasClientPaginationPrevPage;
  };

  return {
    voucherCodes,
    voucherCodesLoading,
    voucherCodesPagination: {
      ...voucherCodesPagination,
      pageInfo: {
        ...voucherCodesPagination.pageInfo,
        hasNextPage: calculateHasNextPage(),
        hasPreviousPage: calculateHasPrevPage(),
      },
      loadNextPage: handleLoadNextPage,
      loadPreviousPage: handleLoadPrevousPage,
    },
    voucherCodesRefetch,
    voucherCodesSettings,
    updateVoucherCodesListSettings: (key: any, value: any) => {
      updateVoucherCodesListSettings(key, value);
      onSettingsChange(key, value);
    },
    selectedVoucherCodesIds,
    handleSetSelectedVoucherCodesIds,
    handleAddVoucherCode,
    handleGenerateMultipeCodes,
  };
};
