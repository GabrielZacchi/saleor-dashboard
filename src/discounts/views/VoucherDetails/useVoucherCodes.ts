import { VoucherCode } from "@dashboard/discounts/components/VoucherCodesDatagrid/types";
import { GenerateMultipleVoucherCodeFormData } from "@dashboard/discounts/components/VoucherCodesGenerateDialog";
import { useVoucherCodesPagination } from "@dashboard/discounts/components/VoucherCreatePage/hooks/useVoucherCodesPagination";
import { generateMultipleIds } from "@dashboard/discounts/components/VoucherCreatePage/utils";
import { useVoucherCodesQuery } from "@dashboard/graphql";
import useListSettings from "@dashboard/hooks/useListSettings";
import useLocalPaginator, {
  LocalPagination,
  useLocalPaginationState,
} from "@dashboard/hooks/useLocalPaginator";
import { useRowSelection } from "@dashboard/hooks/useRowSelection";
import { ListViews } from "@dashboard/types";
import { mapEdgesToItems } from "@dashboard/utils/maps";
import isEqual from "lodash/isEqual";
import { useCallback, useState } from "react";

export const useVoucherCodes = ({ id }: { id: string }) => {
  const {
    settings: voucherCodesSettings,
    updateListSettings: updateVoucherCodesListSettings,
  } = useListSettings(ListViews.VOUCHER_CODES);

  const [addedVoucherCodes, setAddedVoucherCodes] = useState<VoucherCode[]>([]);
  const [isServerPagination, setIsServerPagination] = useState(true);

  const handleAddVoucherCode = (code: string) => [
    setAddedVoucherCodes(codes => [...codes, { code, status: "Draft" }]),
  ];

  const handleGenerateMultipeCodes = ({
    quantity,
    prefix,
  }: GenerateMultipleVoucherCodeFormData) => {
    setAddedVoucherCodes(codes => [
      ...codes,
      ...generateMultipleIds(quantity, prefix),
    ]);
    setIsServerPagination(false);
  };

  const [
    serverVoucherCodesPaginationState,
    setServerVoucherCodesPaginationState,
  ] = useLocalPaginationState(voucherCodesSettings.rowNumber);

  const serverVoucherCodesPaginate = useLocalPaginator(
    setServerVoucherCodesPaginationState,
  );

  const {
    paginatedCodes: clientVoucherCodes,
    pagination: clientVoucherCodesPagination,
    onSettingsChange,
  } = useVoucherCodesPagination(addedVoucherCodes);

  const hasClientPaginationNextPage =
    clientVoucherCodesPagination?.pageInfo?.hasNextPage;
  const hasClientPaginationPrevPage =
    clientVoucherCodesPagination.pageInfo?.hasPreviousPage;

  const freeSlotsInClientPagianationPage =
    voucherCodesSettings.rowNumber - clientVoucherCodes.length;

  const {
    data: voucherCodesData,
    loading: voucherCodesLoading,
    refetch: voucherCodesRefetch,
  } = useVoucherCodesQuery({
    skip: !isServerPagination && hasClientPaginationNextPage,
    variables: {
      id,
      first:
        !isServerPagination &&
        !hasClientPaginationNextPage &&
        freeSlotsInClientPagianationPage > 0
          ? freeSlotsInClientPagianationPage
          : undefined,
      ...(isServerPagination && serverVoucherCodesPaginationState),
    },
  });

  const serverVoucherCodesPagination = serverVoucherCodesPaginate(
    voucherCodesData?.voucher?.codes?.pageInfo,
    serverVoucherCodesPaginationState,
  );

  const hasServerPaginationNextPage =
    serverVoucherCodesPagination?.pageInfo?.hasNextPage;
  const hasServerPaginationPrevPage =
    serverVoucherCodesPagination?.pageInfo?.hasPreviousPage;

  const {
    selectedRowIds,
    setClearDatagridRowSelectionCallback,
    setSelectedRowIds,
  } = useRowSelection();

  const serverVoucherCodes = (mapEdgesToItems(
    voucherCodesData?.voucher?.codes,
  ) ?? []) as VoucherCode[];

  const freeSlotsInServerPagianationPage =
    voucherCodesSettings.rowNumber - serverVoucherCodes.length;

  let voucherCodes: VoucherCode[] = [];

  if (isServerPagination) {
    voucherCodes = [
      ...(freeSlotsInServerPagianationPage > 0 &&
      !serverVoucherCodesPagination.pageInfo?.hasPreviousPage
        ? addedVoucherCodes.slice(-freeSlotsInServerPagianationPage)
        : []),
      ...serverVoucherCodes,
    ];
  } else {
    voucherCodes = [
      ...clientVoucherCodes,
      ...(!hasClientPaginationNextPage && freeSlotsInClientPagianationPage > 0
        ? [...serverVoucherCodes]
        : []),
    ];
  }

  let voucherCodesPagination: LocalPagination;

  if (!isServerPagination) {
    voucherCodesPagination = clientVoucherCodesPagination;
  } else {
    voucherCodesPagination = serverVoucherCodesPagination;
  }

  const handleSetSelectedVoucherCodesIds = useCallback(
    (rows: number[], clearSelection: () => void) => {
      if (!voucherCodes) {
        return;
      }

      const rowsIds = rows.map(row => voucherCodes[row].code).filter(Boolean);
      const haveSaveValues = isEqual(rowsIds, selectedRowIds);

      if (!haveSaveValues) {
        setSelectedRowIds(rowsIds as string[]);
      }

      setClearDatagridRowSelectionCallback(clearSelection);
    },
    [
      voucherCodes,
      selectedRowIds,
      setClearDatagridRowSelectionCallback,
      setSelectedRowIds,
    ],
  );

  return {
    voucherCodes,
    voucherCodesLoading,
    voucherCodesPagination: {
      ...voucherCodesPagination,
      pageInfo: {
        ...voucherCodesPagination.pageInfo,
        hasNextPage: isServerPagination
          ? hasServerPaginationNextPage
          : hasClientPaginationNextPage || hasServerPaginationNextPage,
        hasPreviousPage: !isServerPagination
          ? hasClientPaginationPrevPage
          : hasServerPaginationPrevPage || hasClientPaginationPrevPage,
      },
      loadNextPage: () => {
        if (isServerPagination) {
          voucherCodesPagination.loadNextPage();
        } else {
          if (clientVoucherCodes.length < voucherCodesSettings.rowNumber) {
            setIsServerPagination(true);
            serverVoucherCodesPagination.loadNextPage();
          } else {
            clientVoucherCodesPagination.loadNextPage();
          }
        }
      },
      loadPreviousPage: () => {
        if (!isServerPagination) {
          clientVoucherCodesPagination.loadPreviousPage();
        } else {
          if (!hasServerPaginationPrevPage && addedVoucherCodes.length > 0) {
            clientVoucherCodesPagination.loadPreviousPage();
            setIsServerPagination(false);
          } else {
            voucherCodesPagination.loadPreviousPage();
          }
        }
      },
    },
    voucherCodesRefetch,
    voucherCodesSettings,
    updateVoucherCodesListSettings: (key: any, value: any) => {
      updateVoucherCodesListSettings(key, value);
      onSettingsChange(key, value);
    },
    selectedVoucherCodesIds: selectedRowIds,
    handleSetSelectedVoucherCodesIds,
    handleAddVoucherCode,
    handleGenerateMultipeCodes,
  };
};
