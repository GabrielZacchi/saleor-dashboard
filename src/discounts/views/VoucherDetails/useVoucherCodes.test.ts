import { useVoucherCodesQuery } from "@dashboard/graphql";
import { act } from "@testing-library/react";
import { renderHook } from "@testing-library/react-hooks";

import { useVoucherCodes } from "./useVoucherCodes";

jest.mock("uuid", () => ({ v4: () => "123456789" }));
jest.mock("@dashboard/graphql", () => ({
  useVoucherCodesQuery: jest.fn(() => ({
    data: {
      voucher: {
        codes: {
          edges: [
            {
              node: {
                code: "code 1",
                used: 0,
              },
            },
            {
              node: {
                code: "code 2",
                used: 0,
              },
            },
          ],
          pageInfo: {
            hasNextPage: true,
            hasPreviousPage: false,
            endCursor: "curson",
            startCursor: "cursor",
          },
        },
      },
    },
    loading: false,
    refetch: jest.fn(),
  })),
}));

describe("useVoucherCodes", () => {
  it("should return correct data", () => {
    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    expect(result.current).toMatchSnapshot();
  });

  it("should return manually generated voucher coded", () => {
    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    act(() => {
      result.current.handleAddVoucherCode("code 3");
      result.current.handleAddVoucherCode("code 4");
    });

    expect(result.current.voucherCodes).toEqual([
      { code: "code 3", status: "Draft" },
      { code: "code 4", status: "Draft" },
      { code: "code 1", used: 0 },
      { code: "code 2", used: 0 },
    ]);
  });

  it("should return automatictlly genereted voucher codes", () => {
    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    act(() => {
      result.current.handleGenerateMultipeCodes({
        quantity: "5",
        prefix: "code",
      });
    });

    expect(result.current.voucherCodes).toEqual([
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code 1", used: 0 },
      { code: "code 2", used: 0 },
    ]);
  });

  it("should allow to paginate voucher codes comes from server", () => {
    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    act(() => {
      result.current.updateVoucherCodesListSettings("rowNumber", 2);
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      false,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );

    expect(result.current.voucherCodes).toEqual([
      { code: "code 1", used: 0 },
      { code: "code 2", used: 0 },
    ]);

    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 3",
                  used: 0,
                },
              },
              {
                node: {
                  code: "code 4",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: true,
              endCursor: "curson",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
    });

    expect(result.current.voucherCodes).toEqual([
      { code: "code 3", used: 0 },
      { code: "code 4", used: 0 },
    ]);

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      false,
    );

    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );

    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 1",
                  used: 0,
                },
              },
              {
                node: {
                  code: "code 2",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: true,
              hasPreviousPage: false,
              endCursor: "curson",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    act(() => {
      result.current.voucherCodesPagination.loadPreviousPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodes).toEqual([
      { code: "code 1", used: 0 },
      { code: "code 2", used: 0 },
    ]);
  });

  it("should allow to paginate voucher codes comes from client", () => {
    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: null,
      loading: false,
    }));

    const generatedCodes = [
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
    ];

    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    act(() => {
      result.current.updateVoucherCodesListSettings("rowNumber", 5);
      result.current.handleGenerateMultipeCodes({
        quantity: "10",
        prefix: "code",
      });
      result.current.handleAddVoucherCode("code 1");
      result.current.handleAddVoucherCode("code 2");
    });

    expect(result.current.voucherCodes).toEqual(generatedCodes.slice(5));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );
    expect(result.current.voucherCodes).toEqual(generatedCodes.slice(5));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      undefined,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );
    expect(result.current.voucherCodes).toEqual([
      { code: "code 1", status: "Draft" },
      { code: "code 2", status: "Draft" },
    ]);

    act(() => {
      result.current.voucherCodesPagination.loadPreviousPage();
    });
    act(() => {
      result.current.voucherCodesPagination.loadPreviousPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      false,
    );
    expect(result.current.voucherCodes).toEqual(generatedCodes.slice(5));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
      result.current.updateVoucherCodesListSettings("rowNumber", 20);
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      undefined,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      false,
    );
    expect(result.current.voucherCodes).toEqual([
      ...generatedCodes,
      { code: "code 1", status: "Draft" },
      { code: "code 2", status: "Draft" },
    ]);
  });

  it("should allow to paginate voucher codes comes from client and server", () => {
    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 1",
                  used: 0,
                },
              },
              {
                node: {
                  code: "code 2",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: true,
              endCursor: "curson",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    const { result } = renderHook(() => useVoucherCodes({ id: "1" }));

    const generatedCodes = [
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
      { code: "code-123456789", status: "Draft" },
    ];

    act(() => {
      result.current.updateVoucherCodesListSettings("rowNumber", 2);
      result.current.handleGenerateMultipeCodes({
        quantity: "3",
        prefix: "code",
      });
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      false,
    );
    expect(result.current.voucherCodes).toEqual(generatedCodes.slice(1));

    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 1",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: true,
              hasPreviousPage: false,
              endCursor: "cursor",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
    });

    expect(useVoucherCodesQuery).toHaveBeenCalledWith({
      skip: false,
      variables: { first: 1, id: "1" },
    });
    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );
    expect(result.current.voucherCodes).toEqual([
      ...generatedCodes.slice(2),
      { code: "code 1", used: 0 },
    ]);

    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 2",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: false,
              hasPreviousPage: true,
              endCursor: "cursor",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    act(() => {
      result.current.voucherCodesPagination.loadNextPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      false,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );
    expect(result.current.voucherCodes).toEqual([{ code: "code 2", used: 0 }]);

    (useVoucherCodesQuery as jest.Mock).mockImplementation(() => ({
      data: {
        voucher: {
          codes: {
            edges: [
              {
                node: {
                  code: "code 1",
                  used: 0,
                },
              },
            ],
            pageInfo: {
              hasNextPage: true,
              hasPreviousPage: false,
              endCursor: "cursor",
              startCursor: "cursor",
            },
          },
        },
      },
      loading: false,
    }));

    act(() => {
      result.current.voucherCodesPagination.loadPreviousPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      true,
    );

    expect(result.current.voucherCodes).toEqual([
      ...generatedCodes.slice(2),
      { code: "code 1", used: 0 },
    ]);

    act(() => {
      result.current.voucherCodesPagination.loadPreviousPage();
    });

    expect(result.current.voucherCodesPagination.pageInfo.hasNextPage).toBe(
      true,
    );
    expect(result.current.voucherCodesPagination.pageInfo.hasPreviousPage).toBe(
      false,
    );
    expect(result.current.voucherCodes).toEqual(generatedCodes.slice(1));
  });
});
