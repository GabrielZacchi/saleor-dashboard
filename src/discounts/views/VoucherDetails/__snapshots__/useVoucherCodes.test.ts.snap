// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useVoucherCodes should return correct data 1`] = `
Object {
  "handleAddVoucherCode": [Function],
  "handleGenerateMultipeCodes": [Function],
  "handleSetSelectedVoucherCodesIds": [Function],
  "selectedVoucherCodesIds": Array [],
  "updateVoucherCodesListSettings": [Function],
  "voucherCodes": Array [
    Object {
      "code": "code 1",
      "used": 0,
    },
    Object {
      "code": "code 2",
      "used": 0,
    },
  ],
  "voucherCodesLoading": false,
  "voucherCodesPagination": Object {
    "loadNextPage": [Function],
    "loadPreviousPage": [Function],
    "pageInfo": Object {
      "endCursor": "curson",
      "hasNextPage": true,
      "hasPreviousPage": false,
      "startCursor": "cursor",
    },
    "paginatorType": "click",
  },
  "voucherCodesRefetch": [MockFunction],
  "voucherCodesSettings": Object {
    "rowNumber": 20,
  },
}
`;
